Module 3 / tasks 1 and 2

GeorgiiDermenzhy opened the pull request:
In this pull request I've added ...
Module 3:
task 1 - create sandwich_maker function using pyinputplus
task 2 - create my_splitter function that mimics built in split() method
	  - use pytest to cover task 2 with unix tests
<!-- Please mention all relevant module/task numbers. -->

# Pull Request Checklist

<!-- Please keep this section. It will make the maintainer's life easier. -->

* [x] I've checked code quality with flake8 and pylint tools
* [x] I've checked compliance with Python docstring conventions with pydocstyle tool
* [x] All third party packages are added to requirements.txt

GeorgiiDermenzhy committed:
Module 3 / tasks 1 and 2:
1) in task 1 - create sandwich_maker function using pyinputplus
2) in task 2 - create my_splitter function that mimics built in split() method
			 - use pytest to cover task 2 with unix tests

musk-ox commented on tasks/Module 3/task_2_test.py:
```diff
@@ -0,0 +1,47 @@
+    """Create personal str.split() implementation with UnitTest."""
+    if separator is None:
+        raise TypeError(
+            "YOU SHALL NOT PASS..unless you'll provide a separator")
```
I like the error message, however the task was to:
> implement my_split function that mimics built in split() method as close as possible.

Please take a look at docs for `split` method of `str` type:
```
>>> help(str.split)
```

GeorgiiDermenzhy replied:
error has been replaced with findall method of regex.

@musk-ox @Pacu2 Guys, could you please advice, if we could(somehow) use our `separator` argument as an input for regex? I know we should use raw strings for it, but maybe you know a way? :)

musk-ox commented on tasks/Module 3/task_1.py:
```diff
@@ -0,0 +1,59 @@
+            sandwich_price = SAUCE[sauce_type] + sandwich_price
+            sandwich_ingredients.append(sauce_type)
+
+        number_of_sandwiches = pyip.inputInt(
```
I think it is a good idea to introduce some minimal amount of sandwiches that someone can buy. Otherwise you may get few sandwiches while loosing a lot of gold at the same time :)
```
How many sandwiches would you like to buy?
-2
Your sandwich consists of white, chicken. You wanted -2 sandwich(es). That's a total of -69.60 kg of gold
```
You can add `min=1` to `inputInt`.

GeorgiiDermenzhy replied:
Minumal amount of sandwiches added to inputInt

musk-ox commented on tasks/Module 3/task_1.py:
```diff
@@ -0,0 +1,59 @@
+    sandwich's ingredients;
+    order's price
+    """
+    while True:
```
What is the purpose of this loop? It passes only one time anyway because of the `break` statement at the end so please remove it.

GeorgiiDermenzhy replied:
Loop removed

musk-ox commented on tasks/Module 3/task_1.py:
```diff
@@ -0,0 +1,59 @@
+        sandwich_price = "{:.2f}".format(sandwich_price * number_of_sandwiches)
+
+        print(
+            "Your sandwich consists of " + ', '.join(sandwich_ingredients)
```
If you want you can even put join operation into f-string:
```
print(
    f"Your sandwich consists of {', '.join(sandwich_ingredients)}. " 
    f"You wanted {number_of_sandwiches} sandwich(es). "
    f"That's a total of {sandwich_price} kg of gold")
```
I find it slightly more readable (especially that you leave period in the same line), but it is up to you.

GeorgiiDermenzhy replied:
print statement has been changed, based on advice above

musk-ox commented on tasks/Module 3/task_1.py:
```diff
@@ -0,0 +1,59 @@
+    """
+    while True:
+
+        bread_type = pyip.inputMenu(['wheat', 'white', 'sourdough'])
```
I would extract all choices into constants just like you did with prices. Here you can introduce constant like `BREAD_TYPE_CHOICES` or something similar.

GeorgiiDermenzhy replied:
Hi @Wiktor
first i thought to create new costants for bread etc, but i tought "why not to use what we already have?" and wanted to use all keys in BREAD as BREAD.keys. Unfortunatelly this throwing an error:

`File "E:/Basics-of-Python/tasks/Module 3/task_1.py", line 21, in sandwich_maker
    bread_type = pyip.inputMenu(BREAD.keys())
  File "E:\Basics-of-Python\venv\lib\site-packages\pyinputplus\__init__.py", line 747, in inputMenu
    pysv._validateParamsFor_validateChoice(
  File "E:\Basics-of-Python\venv\lib\site-packages\pysimplevalidate\__init__.py", line 707, in _validateParamsFor_validateChoice
    raise PySimpleValidateException("choices arg must be a sequence")
pysimplevalidate.PySimpleValidateException: choices arg must be a sequence`

is there a way to get key from BREAD dict and use it in inputMenu, or it's better to have 4 more constants for every ingredient

Pacu2 replied:
I think parsing it to the list will help. Like `bread_type = pyip.inputMenu(list(BREAD.keys()))`

musk-ox replied:
@GeorgiiDermenzhy it is actually very good idea to re-use existing constants, much better than introducing new ones 👍 

The error message `pysimplevalidate.PySimpleValidateException: choices arg must be a sequence` may look cryptic at the beginning, but basically sequences are collections`*` that are ordered and reversible: you can access element at arbitrary position by `sequence[some_int]` and you can use well... `reversed` function on it. Built-in sequences in you knowPython are `list`, `tuple` and `str`. Others like `dict` or `set` are collections but not sequences. Both sequences and collections are "iterable" which means you can use them in for loop. 

**Warning: more technical details below**
In addition object returned by `dict.keys()` is not regular Python collection you know but instance of special `dict_keys`. It has special abilities eg. it knows about changes in your dictionary:
```
>>> d = {'a': 1}
>>> k = d.keys()
>>> k
dict_keys(['a'])
>>> d['b'] = 2
>>> k
dict_keys(['a', 'b'])
```
But it is not a sequence:
```
>>> k[0]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'dict_keys' object is not subscriptable
>>> reversed(k)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'dict_keys' object is not reversible
```

`*` You may now wondering what is exact definition of collection :) Basically anything that supports all of following techniques:

- `for something in collection:`
- `len(collection)`
- `something in collection`

So you can do all these operation with all object from this comment, but only with sequences you can do:
- `sequence[3]`
- `reversed(sequence)`

GeorgiiDermenzhy replied:
Hi Guys,
Thanks for the notes :)
@Filip, parse it to list worked perfect.
@Wiktor, I was reading it and re-reading xD don't know, if I got this(most probably not), but per my understanding `InputMenu` is awaiting for a `list/tuple` (that is a sequence) to be provided as an argument, but instead of it, it got `collection` that is not a `sequence`. 
And sequence is required because ....I don't know, could "list" all the keys, as `string` or `set` of strings? So, when I was using `BREAD.keys` it was trying to provide an address of `dict_keys object`. And this `dict_keys` is an object but not a sequence?

musk-ox replied:
@GeorgiiDermenzhy you're almost right. I think the key point here is that if something expects sequence of strings, you can give it either list or tuple.  When you provided `BREAD.keys()`, you provided `dict_keys` object that is collection but not sequence.

GeorgiiDermenzhy replied:
All inputMenu has been changed to use keys of CONSTs

musk-ox commented on tasks/Module 3/task_1.py:
```diff
@@ -0,0 +1,59 @@
+        sandwich_ingredients.append(bread_type)
+
+        protein_type = pyip.inputMenu(['chicken', 'turkey', 'ham', 'tofu'])
+        sandwich_price = PROTEIN[protein_type] + sandwich_price
```
It can be also written using `+=` operator:
```
sandwich_price += PROTEIN[protein_type]
```

GeorgiiDermenzhy replied:
Incrementation changed

musk-ox commented on tasks/Module 3/task_1.py:
```diff
@@ -0,0 +1,59 @@
+
+        number_of_sandwiches = pyip.inputInt(
+            prompt="How many sandwiches would you like to buy?\n")
+        sandwich_price = "{:.2f}".format(sandwich_price * number_of_sandwiches)
```
Formatting of floats can be also done with f-strings although syntax is a bit different:
```
sandwich_price = f"{sandwich_price * number_of_sandwiches:.2f}"
```
Also you are assigning string to `sandwich_price` although it was earlier float. Python allows this due to its dynamic nature but this is usually not the best idea, especially with longer code.

Personally I would introduce new variable `total_price = sandwich_price * number_of_sandwiches` and then format it in f-string in print statement `f"That's a total of {total_price:.2f} kg of gold"`. Of course if you are rich enough to don't care about grams of gold :)

GeorgiiDermenzhy replied:
`total_price` variable introduced

GeorgiiDermenzhy committed:
v.2 for task 1 and 2 of module 2:
task 1: - use list(BREAD.keys()) as an input for inputManu
		- change incrementation of sandwich_price to use shorter version of it
		- introduce minimal amount of sandwiches
		- introduce new variable "total_price" for total sandwiche/s price
task 2: - use regexe's findall method to split string without provided separator

GeorgiiDermenzhy committed:
Update task_1 Module 3:
change all inputMenu to use keys of CONSTs

musk-ox commented on tasks/Module 3/task_2_test.py:
```diff
@@ -0,0 +1,48 @@
+    split_list = []
+    separator_len = len(separator)
+
+    for i in to_split:
```
The PyLint's complains with R1710 error because you have `return` statement only under the condition. To linter it means that your function returns something (list of strings) conditionally and in the other case it implicitly returns `None` and it does not like it. You can take a look at [Python style guide here](https://www.python.org/dev/peps/pep-0008/#programming-recommendations) and search for "Be consistent in return statements"

Of course we are smarter than PyLint and we know that it will never return `None`. Direct solution to this is add explicit `return None` at the end of the function, **but** please don't do it for now, because we can do slightly better by applying some changes described below :)

So, two thoughts regarding this algorithm.

1. The `for` loop may be not the best choice here. It basically says: "for every character in input string do blah blah blah", but your code does not perform operations per character. Instead it works until it reach `else` clause and then returns. It cases where you want to repeat something if condition is met, the better choice is `while` loop.
2. In `else` clause this slicing `to_split[:len(to_split)]` is not necessary, because if `to_split` does not contain separator, then it means that `to_split` is the last word. Actually it is worth noticing that we want to execute this code unconditionally with every function call - we always have this last word left after the last occurrence of separator. 
3. This relates to PyLint issue. We can pull this return outside the loop. In such way, loop will be responsible for collecting words and will end when we have all words on `separated_word` list.

Having in mind points above you can try to rewrite your algorithm using `while` loop. You can start with `while separator in to_split:` or `while to_split:` (you can use strings in conditions just like this - empty string is False, nonempty is True in this context).

GeorgiiDermenzhy replied:
@musk-ox thanks for advices!! Now code looks more simple and readeble
i've replaced `for loop` with `while`, removed main `if condition` and replaced `else` with new `if` that does have reasonable condition

musk-ox commented on tasks/Module 3/task_2_test.py:
```diff
@@ -0,0 +1,48 @@
+
+
+def my_splitter(to_split, separator=None):
+    """Create personal str.split() implementation with UnitTest."""
```
Remember, that docstring should say what your function does. Currently it looks a bit weird, as your function is supposed to create another split function and unit tests for it. I would rather say something like "Split string with provided separator...".

GeorgiiDermenzhy replied:
docstring updated

musk-ox commented on tasks/Module 3/task_2_test.py:
```diff
@@ -0,0 +1,48 @@
+    assert my_splitter(",J", ",") == ["", "J"]
+
+
+def test_without_separator():
```
Maybe it is worth to add one more test with string containing double spaces and and `separator=" "`. What do you thing?

GeorgiiDermenzhy replied:
new `test_double_separator` implemented

GeorgiiDermenzhy committed:
update Module 3 task 2 file with:
1) for loop replaced with while
2) new def test_double_spaces introduced

musk-ox commented on tasks/Module 3/task_2_test.py:
```diff
@@ -44,5 +41,11 @@ def test_two_chars_and_separator():
 
 def test_without_separator():
     """Test function without separator provided."""
+    assert my_splitter("string with !@#$double spaces") == \
```
Why does "string with !@#$double spaces" have singe spaces? :)

I think there should be two case for string with double spaces:

- test if `my_splitter("string  with  !@#$double  spaces and new\nline")` returns `["string", "with", "!@#$double", "spaces", "and", "new", "line"]` - empty strings are filtered out
- test if `my_splitter("string  with  !@#$double  spaces", separator=" ")` returns `["string", "", "with", "", "!@#$double", "", "spaces"]`

musk-ox replied:
Please also add a test case then string ends with separator:
`my_splitter("aaa,bbb,", ",")`
It is interesting case that shows small bug in your algorithm :)

GeorgiiDermenzhy replied:
@musk-ox infinite loop is not a small bug ))
as for the tests, i thought that we need to have two cases;
1) when in first we are not providing a separator and words separated by one space
2) the safe as above, but words separated with two spaces

I've fixed this alread, with below:

`def test_without_separator():`
   ` """Test function without separator provided."""`
    `assert my_splitter("string  with  !@#$double  spaces") == \`
           `["string", "with", "!@#$double", "spaces"]`


`def test_double_spaces():`
    `"""Test function with with string with double spaces and space`
   ` as a separator"""`
    `assert my_splitter("string  with  !@#$double  spaces", " ") == \`
           `["string", "", "with", "", "!@#$double", "", "spaces"]`

musk-ox replied:
> infinite loop is not a small bug

I did't want to cause panic :)

> I've fixed this alread, with below 

Did you push it? I cannot see it here.

GeorgiiDermenzhy replied:
@musk-ox no, i still looking for a solution with `my_splitter("aaa,bbb,", ",")` xD

GeorgiiDermenzhy replied:
new implementation introduced based on above

GeorgiiDermenzhy committed:
update Module 3 task 2 file with:
1) Update separators_location to avoid infinite loop in case, when string ends with separator
2) new def test_string_ends_with_sep introduced

GeorgiiDermenzhy committed:
Move `if separator not in to_split` outside of the loop

GeorgiiDermenzhy committed:
update Module 3 task 2 file with:
Remove unheeded `if` statement after `while` loop
